### PR TITLE
Clean up and simplify path header handling.

### PR DESCRIPTION
**Pull request by kormat**
- Switch to using indexes instead of offsets /and/ indexes to address
  opaque fields.
- lib.packet.path.\* now use a new OpaqueFieldList class, which manages
  the opaque fields for them. It handles indexing, packing, reversing.
  This grealy simplifies the path.py code.
- Consolidated IOF/OF pointer methods in lib.packet.scion to 4 simple
  functions, replacing 9 existing ones.
- Lots of consolidation of path handling methods in path.py.
- Cleanup of path.PathConstructor, including switching everything from
  staticmethod to classmethod, to allow much cleaner self-referencing.
- Rewrote unit tests for path.py to be much cleaner.

Minor stuff:
- Print the actual OF info type name.
- Cleanup of OF str formatting.
- Removed unhelpful **repr** methods that screw up unit testing errors.
- Cleanup of comments and docstrings in path.py.

Everything seems to work, including end2end_test.py.

(Depends on #350 and #351)
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38855329%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38855706%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38856347%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38856468%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38856783%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38856952%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38856968%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38857184%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38857209%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23issuecomment-138282084%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38857859%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38857993%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38858712%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23issuecomment-138282084%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22much%20cleaner%2C%20lgtm%20%21%22%2C%20%22created_at%22%3A%20%222015-09-07T12%3A18%3A41Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%207a418a3892c52fdaf1b62f60775b7d6f5a1a1f6f%20lib/packet/opaque_field.py%2098%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38855706%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Maybe%20I%20don%27t%20understand%20but%20IOF%20without%20following%20HOFs%20is%20incorrect.%20Do%20we%20encapsulate%20IOF%20and%20HOFs%20separately%3F%22%2C%20%22created_at%22%3A%20%222015-09-07T11%3A45%3A42Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22the%20code%20clarified%20that.%20%3A%2B1%3A%20%22%2C%20%22created_at%22%3A%20%222015-09-07T11%3A57%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/opaque_field.py%3AL283-453%22%7D%2C%20%22Pull%207a418a3892c52fdaf1b62f60775b7d6f5a1a1f6f%20lib/packet/scion.py%20154%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38856952%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22btw.%20at%20some%20point%20we%20should%20agree%20on%20when%20to%20use%20%60assert%60%20and%20when%20exceptions.%22%2C%20%22created_at%22%3A%20%222015-09-07T12%3A08%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22We%20should%2C%20this%20particular%20one%20shouldn%27t%20be%20an%20assertion%20though%2C%20so%20i%27ll%20go%20fix%20it%20%3A%29%20%5Cr%5Cn%5Cr%5Cn%28Roughly%2C%20my%20thinking%20is%20that%20exceptions%20should%20be%20for%20errors%20that%20can%20be%20caused%20by%20bad/malicious%20data%2C%20assertions%20for%20things%20that%20only%20code%20bugs%20could%20trigger.%20This%20is%20what%20myself%20and%20%40shitz%20negotiated%20when%20reviewing%20his%20rewritten%20router%20%3A%29%22%2C%20%22created_at%22%3A%20%222015-09-07T12%3A13%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22Fixed.%22%2C%20%22created_at%22%3A%20%222015-09-07T12%3A25%3A35Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/scion.py%3AL379-433%22%7D%2C%20%22Pull%207a418a3892c52fdaf1b62f60775b7d6f5a1a1f6f%20lib/packet/scion.py%2050%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38857209%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22actually%2C%20in%20a%20following%20PR%20we%20should%20pack%20%60_idx%60%20values%20directly%20%28not%20calculated%20byte%20offset%29.%22%2C%20%22created_at%22%3A%20%222015-09-07T12%3A13%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22If%20we%20define%20that%20all%20OFs%20have%20a%20fixed%20size%2C%20this%20works%20great%2C%20and%20simplifes%20things%20nicely.%20Some%20of%20the%20existing%20code%20assumes%20IOFs%20and%20HOFs%20could%20have%20different%20sizes%2C%20and%20some%20assumed%20they%20had%20the%20same.%22%2C%20%22created_at%22%3A%20%222015-09-07T12%3A27%3A46Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22IMO%20it%20is%20enough%20if%20we%20assume%20that%20they%20are%20padded%20to%208B%5Cn%5CnOn%207%20September%202015%20at%2014%3A27%2C%20Stephen%20Shirley%20%3Cnotifications%40github.com%3E%5Cnwrote%3A%5Cn%5Cn%3E%20In%20lib/packet/scion.py%5Cn%3E%20%3Chttps%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38857993%3E%3A%5Cn%3E%5Cn%3E%20%3E%20%20%20%20%20%20%20%20%20%20return%20struct.pack%28%5C%22%21HHBBBB%5C%22%2C%20types%2C%20self.total_len%2C%5Cn%3E%20%3E%20-%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20self.curr_iof_p%2C%20self.curr_of_p%2C%5Cn%3E%20%3E%20%2B%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20curr_iof_p%2C%20curr_of_p%2C%5Cn%3E%5Cn%3E%20If%20we%20define%20that%20all%20OFs%20have%20a%20fixed%20size%2C%20this%20works%20great%2C%20and%5Cn%3E%20simplifes%20things%20nicely.%20Some%20of%20the%20existing%20code%20assumes%20IOFs%20and%20HOFs%5Cn%3E%20could%20have%20different%20sizes%2C%20and%20some%20assumed%20they%20had%20the%20same.%5Cn%3E%5Cn%3E%20%5Cu2014%5Cn%3E%20Reply%20to%20this%20email%20directly%20or%20view%20it%20on%20GitHub%5Cn%3E%20%3Chttps%3A//github.com/netsec-ethz/scion/pull/353/files%23r38857993%3E.%5Cn%3E%5Cn%22%2C%20%22created_at%22%3A%20%222015-09-07T12%3A39%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/scion.py%3AL168-199%22%7D%2C%20%22Pull%207a418a3892c52fdaf1b62f60775b7d6f5a1a1f6f%20lib/packet/path.py%201474%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38856783%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22yes%2C%20i%20think%20it%20is%20possible%22%2C%20%22created_at%22%3A%20%222015-09-07T12%3A06%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok.%20We%27d%20need%20additional%20logic%20to%20handle%20that%20properly%20in%20the%20future%2C%20right%20now%20we%27ll%20just%20end%20up%20one%20randomly.%22%2C%20%22created_at%22%3A%20%222015-09-07T12%3A09%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/path.py%3AL604-828%22%7D%2C%20%22Pull%207a418a3892c52fdaf1b62f60775b7d6f5a1a1f6f%20lib/errors.py%204%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/353%23discussion_r38855329%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22what%27s%20an%20advantage%20of%20own%20classes%20for%20errors%2C%20instead%20of%20using%20just%20KeyError%3F%22%2C%20%22created_at%22%3A%20%222015-09-07T11%3A38%3A55Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%20provides%20more%20information%20about%20the%20context.%20You%20can%20do%20things%20like%3A%5Cr%5Cn%60%60%60%5Cr%5Cntry%3A%5Cr%5Cn%20%20%20thing%5Cr%5Cnexcept%20SCIONBaseError%20as%20e%3A%5Cr%5Cn%20%20%20...%20ok%2C%20we%20raised%20an%20exception%2C%20do%20stuff%20...%5Cr%5Cnexcept%20Exception%3A%5Cr%5Cn%20%20%20...%20this%20is%20really%20bad%2C%20this%20is%20an%20unhandled%20exception%2C%20do%20emergency%20stuff%20...%5Cr%5Cn%60%60%60%22%2C%20%22created_at%22%3A%20%222015-09-07T12%3A00%3A35Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/errors.py%3AL48-61%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull 7a418a3892c52fdaf1b62f60775b7d6f5a1a1f6f lib/errors.py 4'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/353#discussion_r38855329'>File: lib/errors.py:L48-61</a></b>
- <a href='https://github.com/pszalach'></a> what's an advantage of own classes for errors, instead of using just KeyError?
- <a href='https://github.com/kormat'></a> It provides more information about the context. You can do things like:

```
try:
thing
except SCIONBaseError as e:
... ok, we raised an exception, do stuff ...
except Exception:
... this is really bad, this is an unhandled exception, do emergency stuff ...
```
- [ ] <a href='#crh-comment-Pull 7a418a3892c52fdaf1b62f60775b7d6f5a1a1f6f lib/packet/path.py 1474'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/353#discussion_r38856783'>File: lib/packet/path.py:L604-828</a></b>
- <a href='https://github.com/pszalach'></a> yes, i think it is possible
- <a href='https://github.com/kormat'></a> Ok. We'd need additional logic to handle that properly in the future, right now we'll just end up one randomly.
- [ ] <a href='#crh-comment-Pull 7a418a3892c52fdaf1b62f60775b7d6f5a1a1f6f lib/packet/scion.py 154'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/353#discussion_r38856952'>File: lib/packet/scion.py:L379-433</a></b>
- <a href='https://github.com/pszalach'></a> btw. at some point we should agree on when to use `assert` and when exceptions.
- <a href='https://github.com/kormat'></a> We should, this particular one shouldn't be an assertion though, so i'll go fix it :)
  (Roughly, my thinking is that exceptions should be for errors that can be caused by bad/malicious data, assertions for things that only code bugs could trigger. This is what myself and @shitz negotiated when reviewing his rewritten router :)
- <a href='https://github.com/kormat'></a> Fixed.
- [ ] <a href='#crh-comment-Pull 7a418a3892c52fdaf1b62f60775b7d6f5a1a1f6f lib/packet/scion.py 50'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/353#discussion_r38857209'>File: lib/packet/scion.py:L168-199</a></b>
- <a href='https://github.com/pszalach'></a> actually, in a following PR we should pack `_idx` values directly (not calculated byte offset).
- <a href='https://github.com/kormat'></a> If we define that all OFs have a fixed size, this works great, and simplifes things nicely. Some of the existing code assumes IOFs and HOFs could have different sizes, and some assumed they had the same.
- <a href='https://github.com/pszalach'></a> IMO it is enough if we assume that they are padded to 8B
  On 7 September 2015 at 14:27, Stephen Shirley notifications@github.com
  wrote:
  > In lib/packet/scion.py
  > https://github.com/netsec-ethz/scion/pull/353#discussion_r38857993:
  >
  > >          return struct.pack("!HHBBBB", types, self.total_len,
  > > -                           self.curr_iof_p, self.curr_of_p,
  > > +                           curr_iof_p, curr_of_p,
  >
  > If we define that all OFs have a fixed size, this works great, and
  > simplifes things nicely. Some of the existing code assumes IOFs and HOFs
  > could have different sizes, and some assumed they had the same.
  >
  > —
  > Reply to this email directly or view it on GitHub
  > https://github.com/netsec-ethz/scion/pull/353/files#r38857993.
  >
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/353#issuecomment-138282084'>General Comment</a></b>
- <a href='https://github.com/pszalach'></a> much cleaner, lgtm !
- [x] <a href='#crh-comment-Pull 7a418a3892c52fdaf1b62f60775b7d6f5a1a1f6f lib/packet/opaque_field.py 98'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/353#discussion_r38855706'>File: lib/packet/opaque_field.py:L283-453</a></b>
- <a href='https://github.com/pszalach'></a> Maybe I don't understand but IOF without following HOFs is incorrect. Do we encapsulate IOF and HOFs separately?
- <a href='https://github.com/pszalach'></a> the code clarified that. :+1:

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/353?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/353?approve=1'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/353'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: lib/errors.py [hunk 1]
```diff
@@ -48,6 +48,13 @@ class SCIONIndexError(SCIONBaseError):
     pass
 
 
+class SCIONKeyError(SCIONBaseError):
```
Comment (pszal) [inline lib/errors.py:51]: what's an advantage of own classes for errors, instead of using just KeyError?

File: lib/packet/opaque_field.py [hunk 2]
```diff
@@ -270,3 +283,170 @@ def __eq__(self, other):
...
         else:
             return False
+
+
+class OpaqueFieldList(object):
+    """
+    Encapsulates lists of Opaque Fields (OFs).
+
+    The lists are stored under labels, where each label describes the contents
+    of the list. Some label will only ever have 1 entry, such as an up segment
```
Comment (pszal) [inline lib/packet/opaque_field.py:293]: Maybe I don't understand but IOF without following HOFs is incorrect. Do we encapsulate IOF and HOFs separately?

Comment (pszal) [inline lib/packet/opaque_field.py:293]: the code clarified that. :+1:

File: lib/errors.py (hunk 1 above)
Comment (kormat) [inline lib/errors.py:51]: It provides more information about the context. You can do things like:

```
try:
   thing
except SCIONBaseError as e:
   ... ok, we raised an exception, do stuff ...
except Exception:
   ... this is really bad, this is an unhandled exception, do emergency stuff ...
```

File: lib/packet/path.py [hunk 3]
```diff
@@ -740,207 +604,224 @@ def _build_shortcut_path(up_segment, down_segment):
...
+        upstream_hof.info = OpaqueFieldType.NORMAL_OF
+        return iof, hofs, upstream_hof
+
+    @classmethod
+    def _join_shortcuts_peer(cls, up_ad, down_ad):
+        """
+        Finds the peering :any:`HopOpaqueField` of the shortcut path.
         """
+        # FIXME(kormat): Is it possible for there to be multiple matches? Could
+        # 2 ADs have >1 peering link to the other?
```
Comment (pszal) [inline lib/packet/path.py:822]: yes, i think it is possible

File: lib/packet/scion.py [hunk 4]
```diff
@@ -370,88 +379,54 @@ def pack(self):
...
         """
-        Returns (number_of_current_of + n)th opaque field. n may be negative.
-        """
-        if self._path is None:
-            return None
-        offset = (self.common_hdr.curr_of_p - (self.common_hdr.src_addr_len +
-                  self.common_hdr.dst_addr_len))
-        return self._path.get_of(offset // OpaqueField.LEN + n)
+        self.common_hdr.curr_iof_idx = self.common_hdr.curr_of_idx + n
+        assert 0 <= self.common_hdr.curr_iof_idx
```
Comment (pszal) [inline lib/packet/scion.py]: btw. at some point we should agree on when to use `assert` and when exceptions.

File: lib/packet/path.py (hunk 3 above)
Comment (kormat) [inline lib/packet/path.py:822]: Ok. We'd need additional logic to handle that properly in the future, right now we'll just end up one randomly.

File: lib/packet/scion.py (hunk 4 above)
Comment (kormat) [inline lib/packet/scion.py]: We should, this particular one shouldn't be an assertion though, so i'll go fix it :) 

(Roughly, my thinking is that exceptions should be for errors that can be caused by bad/malicious data, assertions for things that only code bugs could trigger. This is what myself and @shitz negotiated when reviewing his rewritten router :)

File: lib/packet/scion.py [hunk 5]
```diff
@@ -167,23 +168,31 @@ def pack(self):
         """
         types = ((self.version << 12) | (self.src_addr_type << 6) |
                  self.dst_addr_type)
+        first_of_offset = self.LEN + self.src_addr_len + self.dst_addr_len
+        curr_iof_p = first_of_offset + (self.curr_iof_idx * OpaqueField.LEN)
+        curr_of_p = first_of_offset + (self.curr_of_idx * OpaqueField.LEN)
         return struct.pack("!HHBBBB", types, self.total_len,
-                           self.curr_iof_p, self.curr_of_p,
+                           curr_iof_p, curr_of_p,
```
Comment (pszal) [inline lib/packet/scion.py:175]: actually, in a following PR we should pack `_idx` values directly (not calculated byte offset).

File: lib/packet/scion.py (hunk 4 above)
Comment (kormat) [inline lib/packet/scion.py]: Fixed.

File: lib/packet/scion.py (hunk 5 above)
Comment (kormat) [inline lib/packet/scion.py:175]: If we define that all OFs have a fixed size, this works great, and simplifes things nicely. Some of the existing code assumes IOFs and HOFs could have different sizes, and some assumed they had the same.

Comment (pszal) [inline lib/packet/scion.py:175]: IMO it is enough if we assume that they are padded to 8B

**Pull request merged**